Sprint 07

AndreiFerradesign opened the pull request:
Выполнены задачи по седьмому спринту.

AndreiFerradesign committed:
Написаны Unit тесты

AndreiFerradesign committed:
Написаны UI тесты

AndreiFerradesign committed:
Вынес метод convert в Presenter

AndreiFerradesign committed:
Перенес логику кнопок в Presenter

AndreiFerradesign committed:
Проект переведен на паттерн MVP. Добавлен тест для Presenter. Приложение работает, все написанные тесты проходят проверку.

mmbabaev commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,133 @@
+           self.viewController?.hideAnswerBorder()
+           self.viewController?.drawLoader(isShown: false)
+           self.showNextQuestionOrResults()
+           }
```
скачет форматирование. Предлагаю использовать `cmd+A` + `ctrl+I`

mmbabaev commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -64,111 +56,45 @@ final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate {
         
-    private func hideAnswerBorder() {
+    func hideAnswerBorder() {
             imageView.layer.borderWidth = 8
```
повторяется строка в show и hide

mmbabaev commented on MovieQuiz/Services/MoviesLoader.swift:
```diff
@@ -13,8 +13,11 @@ protocol MoviesLoading {
-    private let networkClient = NetworkClient()
+    private let networkClient: NetworkRouting
     
+    init(networkClient: NetworkRouting = NetworkClient()) {
```
понравилось, что так сделали

mmbabaev commented on MovieQuizPresenterTests/MovieQuizPresenterTests.swift:
```diff
@@ -0,0 +1,57 @@
+            
+            let emptyData = Data()
+            let question = QuizQuestion(image: emptyData, text: "Question Text", correctAnswer: true)
+            let viewModel = sut.convert(model: question)
```
Если есть желание, предлагаю немного улучшить, уже создан `MovieQuizViewControllerProtocolMock`: 

1) Вообще функция convert может быть приватной, она не используется в vc
2) Тогда нужно можно проверить ее косвенно. Проверим метод `didReceiveNextQuestion` 
3) Нужно проверить, что vc получил ожидаемую viewModel. Для этого в моке `func show(quiz step: QuizStepViewModel)` нужно сохранять step в отдельное свойство
4) В тесте вызываем `didReceiveNextQuestion`, затем достаем `viewModel` из мок vc, проверяем его

mmbabaev commented on MovieQuizTests/MoviesLoaderTests.swift:
```diff
@@ -0,0 +1,54 @@
+            // Then
+            switch result {
+            case .success(_):
+                // сравниваем данные с тем, что мы предполагали
```
тут лучше проверять все-таки не просто успешный результат, а что пришел массив, свойства его элементов таки, как мы ожидаем

mmbabaev commented on MovieQuizTests/MoviesLoaderTests.swift:
```diff
@@ -0,0 +1,54 @@
+                // Then
+                switch result {
+                case .failure(let error):
+                    XCTAssertNotNil(error)
```
аналогично здесь лучше проверить что error это именно TestError.test

mmbabaev commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,84 @@
+        let secondPoster = app.images["Poster"] // ещё раз находим постер
+        let indexLabel = app.staticTexts["Index"]
+        
+        sleep(3)
```
необязательно: 
очень много долгих sleep, вообще в production тестах лучше его не использовать, потому что тестов может быть много и они займут большое количество времени. Можно попробовать уменьшить их продолжительность

mmbabaev commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,84 @@
+    
+    func testYesButton() {
+        
+        let firstPoster = app.images["Poster"] // находим первоначальный постер
```
рекомендую удалять учебные комментарии

mmbabaev commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,84 @@
+        sleep(3)
+        
+        XCTAssertTrue(indexLabel.label == "2/10")
+        XCTAssertFalse(firstPoster == secondPoster) // проверяем, что постеры разные
```
Необязательно для исправления, есть неточность в материалах (уже исправляем).

Здесь мы сравниваем XCUIElement XCTAssertFalse(firstPoster == secondPoster)  , но так делать неправильно.
При получении XCUIElement из query скорее всего создается новый экземпляр элемента. Поэтому, если не нажимать на кнопку да, то тест все равно пройдет успешно.

на скрине видно, что адреса объектов разные

Здесь описан вариант решения

https://askcodes.net/coding/xcuielement---obtain-image-value


![sleep(2)](https://user-images.githubusercontent.com/12600807/202518383-e4700c43-4985-4e55-ae41-48130e4d1baf.png)

AndreiFerradesign committed:
Исправлены замечания Ревьюера, а именно:
1. Добавлен тест проверки скрытия Алерта и проверка начала игры по индексу вопроса.
2. Скорректирован UI тест проверки кнопок Yes и No, согласно рекомендации, тест работает, проверку проходит. В случае замены на картинки эквивалентны, тест проваливается.
3. Убрано дублирование кода в функциях showAnswerBorder и hideAnswerBorder
4. Скорректирован Unit тест MoviesLoader
5. Удалены учебные комментарии по всему проекту
6. Уменьшены временные задержки в тестах

mmbabaev commented on MovieQuiz/Services/NetworkClient.swift:
```diff
@@ -22,75 +22,20 @@ struct NetworkClient: NetworkRouting {
     }
 }
-
-struct StubNetworkClient: NetworkRouting {
```
это используется и потерялось

AndreiFerradesign committed:
Вернул на место, извиняюсь если что. Но у меня все компилировалось и запускалось. Надеюсь в этот раз все нормально.
